src/NuGetGallery/Authentication/Providers/CommonAuth/CommonAuthAuthenticator.cs

```
@@ -0,0 +1,145 @@
+            var options = new OpenIdConnectAuthenticationOptions(BaseConfig.AuthenticationType)
+            {
+                RedirectUri = siteRoot + "users/account/authenticate/return",
+                PostLogoutRedirectUri = siteRoot,
+                Scope = OpenIdConnectScopes.OpenIdProfile + " email",
+                ResponseType = OpenIdConnectResponseTypes.CodeIdToken,
+                TokenValidationParameters = new System.IdentityModel.Tokens.TokenValidationParameters() { ValidateIssuer = false },
+            };
+
+            Config.ApplyToOwinSecurityOptions(options);
+
+            app.UseOpenIdConnectAuthentication(options);
+        }
+        
+        public override AuthenticatorUI GetUI()
+        {
+            return new AuthenticatorUI(
+                Strings.MicrosoftAccount_SignInMessage,
+                Strings.MicrosoftAccount_SignInMessage,
+                Strings.MicrosoftAccount_AccountNoun)
+            {
+                IconImagePath = "~/Content/gallery/img/microsoft-account.svg",
+                IconImageFallbackPath = "~/Content/gallery/img/microsoft-account-24x24.png",
+            };
+        }
+
+        public override ActionResult Challenge(string redirectUrl)
+        {
+            return new ChallengeResult(BaseConfig.AuthenticationType, redirectUrl);
+        }
+
+        public override bool IsAuthorForIdentity(ClaimsIdentity claimsIdentity)
+        {
+            Claim issuer = claimsIdentity.FindFirst(V2Claims.Issuer);
+            Claim tenant = claimsIdentity.FindFirst(V2Claims.TenantId);
+            if (issuer == null || tenant == null)
+            {
+                return false;
+            }
+
+            var expectedIssuer = string.Format(Authority, tenant.Value);
+            return string.Equals(issuer.Value, expectedIssuer, StringComparison.OrdinalIgnoreCase);
+        }
+
+        public override AuthInformation GetAuthInformation(ClaimsIdentity claimsIdentity)
```

May be combine this and above method as `TryGet`

---

I looked into this, and I think its better to throw the exception here, given that we want to disallow missing emails going forward. In subsequent PRs I will clean it up to throw a well defined exception that we can surface to the user as missing email.